mlir/include/mlir/Dialect/SCF/IR/SCFOps.td

```diff
@@ -931,6 +931,28 @@ def ReduceReturnOp :
   let hasVerifier = 1;
 }
 
+def ScopeOp : SCF_Op<"scope",
+    [AutomaticAllocationScope,
+     RecursiveMemoryEffects,
+     IsolatedFromAbove,
+     SingleBlockImplicitTerminator<"scf::YieldOp">]> {
+  let summary = "isolated code scope";
+  let description = [{
+    This op models an executed-once region, similar to `execute_region`
+    except it's isolated from above, which facilitates local scoping of
+    code without resorting to function outlining.
```

Comment:
I don't understand the description, and this the motivation for adding this operation right now, can you elaborate?

Also please split the Pr in two: introduce the op to SCF is orthogonal to the Transform dialect.

Reply:
Sure, some background:
This patch is a draft following a C4ML'24 discussion regarding the proposed transform, which facilitates trying transformation sequences on a limited scope without resorting to function outlining or cloning large portions of the IR using transform.alternatives. This transform requires an isolated-from-above op which can temporarily encapsulate the to-be-transformed part of the code.
The scope op was put in the scf dialect mainly since the transform dialect doesn't allow adding non-transform ops (perhaps it should?). However, a EuroLLVM'24 [round table](https://discourse.llvm.org/t/notes-from-the-mlir-upstream-round-table-eurollvm-2024/78374) brought up potential other use cases for such an op, so I'll split it into its own patch to keep the discussions separate.